src/TpetraLinearSystem.C

```diff
@@ -1834,59 +1835,47 @@ void TpetraLinearSystem::copy_tpetra_to_stk(
-  const LinSys::ConstOneDVector & tpetraVector = tpetraField->get1dView();
+  const auto deviceVector = tpetraField->getLocalView<sierra::nalu::DeviceSpace>();
 
-  const unsigned p_rank = bulkData.parallel_rank();
+  const int maxOwnedRowId = maxOwnedRowId_;
+  const unsigned numDof = numDof_;
+  auto entityToLID = entityToLID_;
 
   const stk::mesh::Selector selector = stk::mesh::selectField(*stkField)
     & metaData.locally_owned_part()
     & !(stk::mesh::selectUnion(realm_.get_slave_part_vector()))
     & !(realm_.get_inactive_selector());
 
-  stk::mesh::BucketVector const& buckets =
-    realm_.get_buckets(stk::topology::NODE_RANK, selector);
-
-  for (size_t ib=0; ib < buckets.size(); ++ib) {
-    stk::mesh::Bucket & b = *buckets[ib];
-
-    const unsigned fieldSize = field_bytes_per_entity(*stkField, b) / sizeof(double);
-    ThrowRequire(fieldSize == numDof_);
-
-    const stk::mesh::Bucket::size_type length = b.size();
-    double * stkFieldPtr = (double*)stk::mesh::field_data(*stkField, *b.begin());
-    const stk::mesh::EntityId *naluGlobalId = stk::mesh::field_data(*realm_.naluGlobalId_, *b.begin());
-    for (stk::mesh::Bucket::size_type k = 0 ; k < length ; ++k ) {
-      stk::mesh::Entity node = b[k];
-      const LocalOrdinal localIdOffset = entityToLID_[node.local_offset()];
-      for(unsigned d=0; d < fieldSize; ++d) {
-        const LocalOrdinal localId = localIdOffset + d;
-        bool useOwned = true;
-        LocalOrdinal actualLocalId = localId;
-        if(localId >= maxOwnedRowId_) {
-          actualLocalId = localId - maxOwnedRowId_;
-          useOwned = false;
-        }
+  NGPDoubleFieldType ngpField = realm_.ngp_field_manager().get_field<double>(stkField->mesh_meta_data_ordinal());
+  NGPGlobalIdFieldType ngpNaluIdField = realm_.ngp_field_manager().get_field<stk::mesh::EntityId>(realm_.naluGlobalId_->mesh_meta_data_ordinal());
 
-        if (!useOwned) {
-          stk::mesh::EntityId naluId = naluGlobalId[k];
-          stk::mesh::EntityId stkId = bulkData.identifier(node);
-          std::cout << "P[" << p_rank << "] useOwned = " << useOwned << " localId = " << localId << " maxOwnedRowId_= " << maxOwnedRowId_ << " actualLocalId= " << actualLocalId
-                    << " naluGlobalId= " << naluGlobalId[k] << " stkId= " << stkId << " naluId= " << naluId << std::endl;
-        }
-        ThrowRequire(useOwned);
+  ngp::Mesh ngpMesh = realm_.ngp_mesh();
 
-        const size_t stkIndex = k*numDof_ + d;
-        if (useOwned){
-          stkFieldPtr[stkIndex] = tpetraVector[localId];
+  nalu_ngp::run_entity_algorithm(ngpMesh, stk::topology::NODE_RANK, selector,
+  KOKKOS_LAMBDA(const MeshIndex& meshIdx)
+  {
+      stk::mesh::Entity node = (*meshIdx.bucket)[meshIdx.bucketOrd];
+      const LocalOrdinal localIdOffset = entityToLID[node.local_offset()];
+      stk::mesh::EntityId naluId = ngpNaluIdField.get(meshIdx, 0);
+      stk::mesh::EntityId stkId = ngpMesh.identifier(node);
+      if (stkId == naluId) {
```

Comment:
@alanw0 Is this check necessary? The _selector_ is explicitly excluding image nodes when we have periodic BCs.